src/geometry/euclidean/congruence.lean

```diff
@@ -0,0 +1,303 @@
+λ a, set.range_factorization v₂ $ set.range_splitting v₁ a
+
+-- map does indeed preserve corresponding points
+lemma map_sound (h : v₁ ≅ v₂) (i : ι) : ↑(h.map (set.range_factorization v₁ i)) = v₂ i :=
+begin
+  unfold congruence.map,
+  rw set.range_factorization_coe v₂,
+  rw ← dist_eq_zero,
+  rw ← h,
+  rw dist_eq_zero,
+  rw set.apply_range_splitting v₁,
+  rw set.range_factorization_coe v₁,
+end
+
+-- .map and .symm.map are inverses to eachother
+protected lemma map_inverse_self (h : v₁ ≅ v₂) : function.left_inverse h.symm.map h.map :=
+begin
+  intro x,
+  rw subtype.ext_iff,
+  rw ← set.apply_range_splitting v₁ x,
+  apply h.symm.map_sound (set.range_splitting v₁ x),
+end
+
+-- map as an `equiv`
+protected def equiv (h : v₁ ≅ v₂) : set.range v₁ ≃ set.range v₂ :=
+{ to_fun := h.map,
+  inv_fun := h.symm.map,
+  left_inv := h.map_inverse_self,
+  right_inv := h.symm.map_inverse_self}
+
+lemma equiv_sound (h : v₁ ≅ v₂) (i : ι) : ↑(h.equiv (set.range_factorization v₁ i)) = v₂ i :=
+h.map_sound i
+
+-- map as an `isometry_equiv`
+protected def isometry (h : v₁ ≅ v₂) : set.range v₁ ≃ᵢ set.range v₂ :=
+{ to_equiv := h.equiv,
+  isometry_to_fun :=
+  begin
+    intros fx fy,
+    rw [edist_dist, edist_dist], apply congr_arg ennreal.of_real,
+    rw subtype.dist_eq fx fy,
+    rw [← set.apply_range_splitting v₁ fx, ← set.apply_range_splitting v₁ fy],
+    rw h.to_dist_eq_dist,
+    refl,
+  end}
+
+lemma isometry_sound (h : v₁ ≅ v₂) (i : ι) :
+  ↑(h.isometry (set.range_factorization v₁ i)) = v₂ i :=
+h.map_sound i
+
+
+
+
+
+variables {V₁ V₂ : Type*} [inner_product_space ℝ  V₁] [inner_product_space ℝ V₂]
+  [normed_add_torsor V₁ P₁] [normed_add_torsor V₂ P₂]
+
+
+
+
+
+include V₁ V₂
+
+lemma to_angle_eq_angle (h : v₁ ≅ v₂) (i₁ i₂ i₃ : ι) :
+  ∠ (v₁ i₁) (v₁ i₂) (v₁ i₃) = ∠ (v₂ i₁) (v₂ i₂) (v₂ i₃) :=
+begin
+  unfold euclidean_geometry.angle, unfold inner_product_geometry.angle,
+  have key := abs_le.1 (abs_real_inner_div_norm_mul_norm_le_one _ _),
+  have key':= abs_le.1 (abs_real_inner_div_norm_mul_norm_le_one _ _),
+  rw real.arccos_inj key.1 key.2 key'.1 key'.2,
+  simp [real_inner_eq_norm_mul_self_add_norm_mul_self_sub_norm_sub_mul_self_div_two,
+    ← normed_add_torsor.dist_eq_norm', h.to_dist_eq_dist],
+end
+
+omit V₁ V₂
+
+variables {a b c : P₁} {d e f : P₂}
+
+
+lemma _root_.fin.fin_3_swap_left {α : Type*} (a b c : α) : ![a,b,c] = ![b,a,c] ∘ ![1,0,2] :=
+by {ext x, fin_cases x; refl}
+lemma _root_.fin.fin_3_swap_right {α : Type*} (a b c : α) : ![a,b,c] = ![a,c,b] ∘ ![0,2,1] :=
+by {ext x, fin_cases x; refl}
+
+
+lemma triangle_swap_left (h : ![a,b,c] ≅ ![d,e,f]) : ![b,a,c] ≅ ![e,d,f] :=
+begin
+  rw [fin.fin_3_swap_left b a c, fin.fin_3_swap_left e d f],
+  apply sub_congruence ![1,0,2] h,
+end
+lemma triangle_swap_right (h : ![a,b,c] ≅ ![d,e,f]) : ![a,c,b] ≅ ![d,f,e] :=
+begin
+  rw [fin.fin_3_swap_right a c b, fin.fin_3_swap_right d f e],
+  apply sub_congruence ![0,2,1] h,
+end
```

Comment:
Rather than having such specific lemmas for three points, I think you should have an `iff` version of `sub_congruence` in the case of mapping by an equivalence (that is, a term of a type with an `equiv_class` instance) (marked as a `@[simp]` lemma, so with the more complicated expression on the left of the `iff`). Then, such rewriting should be achieved by using concrete equivalences (such as `equiv.swap`) to carry out the required remapping of indices (and any use of `ext` and `fin_cases` inlined where needed).